fwdctl/src/dump.rs

```diff
@@ -24,50 +30,277 @@ pub struct Options {
+    )]
+    pub stop_key: Option<Key>,
+
+    /// The key to start dumping from (if no key is provided, start from the beginning) in hex format.
+    /// Defaults to None.
+    #[arg(
+        long,
+        required = false,
+        conflicts_with = "start_key",
+        value_name = "START_KEY_HEX",
+        value_parser = key_parser_hex,
+        help = "Start dumping from this key (inclusive) in hex format. Conflicts with start_key"
+    )]
+    pub start_key_hex: Option<Key>,
+
+    /// The key to stop dumping to (if no key is provided, stop to the end) in hex format.
+    /// Defaults to None.
+    #[arg(
+        long,
+        required = false,
+        conflicts_with = "stop_key",
+        value_name = "STOP_KEY_HEX",
+        value_parser = key_parser_hex,
+        help = "Stop dumping to this key (inclusive) in hex format. Conflicts with stop_key"
+    )]
+    pub stop_key_hex: Option<Key>,
+
+    /// The max number of the keys going to be dumped.
+    /// Defaults to None.
+    #[arg(
+        short = 'm',
+        long,
+        required = false,
+        value_name = "MAX_KEY_COUNT",
+        help = "Maximum number of keys going to be dumped."
+    )]
+    pub max_key_count: Option<u32>,
+
+    /// The output format of database dump.
+    /// Possible Values: ["csv", "json", "stdout"].
+    /// Defaults to "stdout"
+    #[arg(
+        short = 'o',
+        long,
+        required = false,
+        value_name = "OUTPUT_FORMAT",
+        value_parser = ["csv", "json", "stdout"],
+        default_value = "stdout",
+        help = "Output format of database dump, default to stdout. CSV and JSON formats are available."
+    )]
+    pub output_format: String,
+
+    /// The output file name of database dump.
+    /// Output format must be set when the file name is set.
+    #[arg(
+        short = 'f',
+        long,
+        requires = "output_format",
+        value_name = "OUTPUT_FILE_NAME",
+        default_value = "dump",
+        help = "Output file name of database dump, default to dump. Output format must be set when the file name is set."
+    )]
+    pub output_file_name: PathBuf,
+
     #[arg(short = 'x', long, help = "Print the keys and values in hex format.")]
     pub hex: bool,
 }
 
 pub(super) async fn run(opts: &Options) -> Result<(), api::Error> {
     log::debug!("dump database {:?}", opts);
-    let cfg = DbConfig::builder().truncate(false);
 
+    let cfg = DbConfig::builder().truncate(false);
     let db = Db::new(opts.db.clone(), cfg.build()).await?;
     let latest_hash = db.root_hash().await?;
     let Some(latest_hash) = latest_hash else {
         println!("Database is empty");
         return Ok(());
     };
     let latest_rev = db.revision(latest_hash).await?;
-    let start_key = opts.start_key.clone().unwrap_or(Box::new([]));
+
+    let start_key = opts
+        .start_key
+        .clone()
+        .or(opts.start_key_hex.clone())
+        .unwrap_or_default();
```

Comment:
nit: the clones here don't seem to be required
```suggestion
let start_key: Box<[u8]> = opts
        .start_key
        .or(opts.start_key_hex)
        .unwrap_or_default();
```

Reply:
Well, not quite, but you could use a &Box<[u8]> and just have start_key point to the right thing. You'd have to set up an empty thing to point to for the default that outlives it though, which is a lot of work just to avoid a single clone at the start of the program.

Reply:
yeah, I tried and found out it is not easy